core/os/device/remotessh/device.go

```diff
@@ -0,0 +1,216 @@
+	WriteFile(ctx context.Context, contents io.Reader, mode string, destPath string) error
+	// ForwardPort forwards the remote port. It automatically selects an open
+	// local port.
+	ForwardPort(ctx context.Context, remoteport int) (int, error)
+}
+
+// binding represents an attached SSH client.
+type binding struct {
+	bind.Simple
+
+	connection    *ssh.Client
+	configuration *Configuration
+	// We duplicate OS here because we need to use it
+	// before we get the rest of the information
+	os device.OSKind
+}
+
+var _ Device = &binding{}
+
+func (binding) Forward(ctx context.Context, local, device uint16) error {
+	return nil
+}
+
+func (binding) RemoveForward(ctx context.Context, local uint16) error {
+	return nil
+}
+
+// Devices returns the list of reachable SSH devices.
+func Devices(ctx context.Context, configuration io.Reader) ([]bind.Device, error) {
+	configurations, err := ReadConfiguration(configuration)
+	if err != nil {
+		return nil, err
+	}
+	devices := make([]bind.Device, 0, len(configurations))
+
+	for _, cfg := range configurations {
+		if device, err := GetConnectedDevice(ctx, cfg); err == nil {
+			devices = append(devices, device)
+		}
+	}
+
+	return devices, nil
+}
+
+func GetSSHAgent() ssh.AuthMethod {
+	if sshAgent, err := net.Dial("unix", os.Getenv("SSH_AUTH_SOCK")); err == nil {
+		return ssh.PublicKeysCallback(agent.NewClient(sshAgent).Signers)
+	}
+	return nil
+}
+
+func GetPrivateKeyAuth(path string) (ssh.AuthMethod, error) {
+	bytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := ssh.ParsePrivateKey(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return ssh.PublicKeys(signer), nil
+}
+
+// GetConnectedDevice returns a device that has an option connection.
+func GetConnectedDevice(ctx context.Context, c Configuration) (Device, error) {
+	auths := []ssh.AuthMethod{}
+	if agent := GetSSHAgent(); agent != nil {
+		auths = append(auths, agent)
+	}
+
+	if c.Keyfile != "" {
+		if auth, err := GetPrivateKeyAuth(c.Keyfile); err == nil {
+			auths = append(auths, auth)
+		}
+	}
+
+	if len(auths) == 0 {
+		return nil, fmt.Errorf("No valid authentication method for SSH connection %s", c.Name)
+	}
+
+	hosts, err := knownhosts.New(c.KnownHosts)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read known hosts %v", err)
+	}
+
+	sshConfig := &ssh.ClientConfig{
+		User:            c.User,
+		Auth:            auths,
+		HostKeyCallback: hosts,
+	}
+
+	connection, err := ssh.Dial("tcp", fmt.Sprintf("%s:%d", c.Host, c.Port), sshConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &binding{
+		connection:    connection,
+		configuration: &c,
+		Simple: bind.Simple{
+			To: &device.Instance{
+				Serial:        "",
+				Configuration: &device.Configuration{},
+			},
+			LastStatus: bind.Status_Online,
+		},
+	}
+
+	kind := device.UnknownOS
+
+	// Try to get the OS string for Mac/Linux
+	if osName, err := b.Shell("uname", "-a").Call(ctx); err == nil {
```

Comment:
Somewhat disappointing there's no API to to get info about the remote OS.
What about windows?

Reply:
For windows we can call "ver" which will return the windows version.